File: src/moddar/domain/account.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from moddar.domain.link import Link


class Account:
    """Represents a Reddit account (user)."""

    def __init__(
        self,
        id: str,
        username: str,
        links: list[Link] | None = None,
    ) -> None:
        self._id = id
        self._username = username
        self._links: list[Link] = links if links is not None else []

    @property
    def id(self) -> str:
        return self._id

    @property
    def username(self) -> str:
        return self._username

    @property
    def links(self) -> list[Link]:
        return self._links

    def add_link(self, link: Link) -> None:
        """Add a Link to this account's list of links."""
        self._links.append(link)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Account):
            return NotImplemented
        return self._id == other._id

    def __repr__(self) -> str:
        return (
            f"Account(id={self._id!r}, username={self._username!r}, "
            f"links={self._links!r})"
        )

File: src/server/domain/account.ts
import type { Link } from "./link.ts";

/** Represents a Reddit account (user). */
export class Account {
  private readonly _id: string;
  private readonly _username: string;
  private readonly _links: Link[];

  constructor({
    id,
    username,
    links = null,
  }: {
    id: string;
    username: string;
    links?: readonly Link[] | null;
  }) {
    this._id = id;
    this._username = username;
    this._links = links ? [...links] : [];
  }

  get id(): string {
    return this._id;
  }

  get username(): string {
    return this._username;
  }

  get links(): readonly Link[] {
    return this._links;
  }

  /** Add a Link to this account's list of links. */
  addLink(link: Link): void {
    this._links.push(link);
  }

  /** Identity-based equality: compares by id only. */
  equals(other: unknown): boolean {
    if (!(other instanceof Account)) {
      return false;
    }
    return this._id === other._id;
  }

  toString(): string {
    return (
      `Account(id=${JSON.stringify(this._id)}, ` +
      `username=${JSON.stringify(this._username)}, ` +
      `links=[${this._links.map((l) => l.toString()).join(", ")}])`
    );
  }
}
